### Pair VII
# ANOVA

In [4]:
# Tratamiento de datos
# -----------------------------------------------------------------------
import numpy as np
import pandas as pd

# Estadísticos
# ------------------------------------------------------------------------------
import statsmodels.api as sm
from statsmodels.formula.api import ols
from statsmodels.multivariate.manova import MANOVA
from sklearn.preprocessing import StandardScaler

En el pair programming de hoy usaremos el set de datos que guardastéis en el pair programming de normalización y estandarización.
Hasta ahora habéis estado evaluando las características de vuestro set de datos y habéis hecho una gran exploración, es el momento de hacer vuestro primer ANOVA! En el ejercicio de hoy tendréis que hacer un ANOVA con vuestro datos y hacer una interpretación de los resultados.
📌 NOTA Puede que vuestros datos no se ajusten o no cumplan todas las asunciones, no pasa nada, haced el ANOVA e interpretad los resultados. En próximas lecciones aprenderemos que podemos hacer cuando nos encontramos en esta situación.

In [5]:
df = pd.read_pickle("../data/6_estandarizacion.pkl")
df.sample(5)

cesta_compra  bebidas_super  bebidas_rest  comida_rest      ropa  \
1899     -0.962211      -0.170173     -0.578638    -0.877031 -1.470777   
2613      0.516738       0.627473      0.911026     1.147412  1.070102   
2099     -0.599526      -0.649762     -0.616133    -0.582709 -0.545831   
656       0.121673      -0.380040      0.862383    -0.281346 -0.531529   
1657     -0.358797      -0.473203      0.009120    -0.311724 -0.329697   

      deportes      cine  marlboro  educacion  transporte  gasolina  \
1899 -1.303008 -0.932619 -0.272332  -0.521286   -0.995692  0.159259   
2613  0.198516  0.797856  0.782231   3.453369    0.457992 -0.137037   
2099 -0.560822 -0.637060 -0.507967  -0.414100   -0.732012  0.640741   
656  -0.351375  0.554364  0.676388  -0.233352    0.010289 -0.192593   
1657  0.016242 -0.431853 -0.440367  -0.217196   -0.408186  0.066667   

      vehiculos  internet     movil  gastos_fijos  alquiler        m2  \
1899   0.160732 -0.659366 -0.850000     -0.979830 -0.675622 -0.668763   
2613   0.211624  0.209413 -0.433333     -0.292259  0.837672  0.408762   
2099  -0.797232 -0.533354 -0.350000     -0.607591 -0.551229 -0.794018   
656    0.155115  0.575149  1.133333      0.105734 -0.373435  0.227704   
1657  -0.720435 -0.095827 -0.350000     -0.531774 -0.196502  0.186889   

      inte_fij_hipo  salario_box  
1899       0.861818    -1.350449  
2613      -0.169091     0.652589  
2099      -0.260000    -0.811417  
656        1.104727     0.396691  
1657      -0.103636    -0.285724

In [6]:
lm = ols('salario_box ~ cesta_compra + bebidas_super + bebidas_rest + comida_rest  + ropa + deportes + cine + marlboro + educacion + transporte + gasolina + vehiculos + internet + movil + gastos_fijos + alquiler + m2 + inte_fij_hipo', data=df).fit()
sm.stats.anova_lm(lm)

df       sum_sq      mean_sq             F         PR(>F)
cesta_compra      1.0  1042.356116  1042.356116  10422.918392   0.000000e+00
bebidas_super     1.0    28.359541    28.359541    283.577918   6.910426e-62
bebidas_rest      1.0   229.984421   229.984421   2299.702396   0.000000e+00
comida_rest       1.0    49.596114    49.596114    495.930558  1.088079e-104
ropa              1.0     0.006067     0.006067      0.060667   8.054556e-01
deportes          1.0     3.057587     3.057587     30.573983   3.382004e-08
cine              1.0    83.540378    83.540378    835.352268  1.256066e-169
marlboro          1.0    11.230371    11.230371    112.296785   5.898178e-26
educacion         1.0     9.144676     9.144676     91.441117   1.775263e-21
transporte        1.0     4.245364     4.245364     42.451019   7.982242e-11
gasolina          1.0     5.283854     5.283854     52.835282   4.206196e-13
vehiculos         1.0     2.951919     2.951919     29.517372   5.812375e-08
internet          1.0     1.750621     1.750621     17.505126   2.915943e-05
movil             1.0     0.089600     0.089600      0.895942   3.439193e-01
gastos_fijos      1.0     3.592898     3.592898     35.926767   2.196820e-09
alquiler          1.0    14.793282    14.793282    147.923698   1.509832e-33
m2                1.0     0.348889     0.348889      3.488676   6.185018e-02
inte_fij_hipo     1.0    14.310772    14.310772    143.098890   1.595992e-32
Residual       4855.0   485.529940     0.100006           NaN            NaN

In [7]:
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:            salario_box   R-squared:                       0.756
Model:                            OLS   Adj. R-squared:                  0.755
Method:                 Least Squares   F-statistic:                     835.9
Date:                Tue, 12 Sep 2023   Prob (F-statistic):               0.00
Time:                        23:16:38   Log-Likelihood:                -1295.1
No. Observations:                4874   AIC:                             2628.
Df Residuals:                    4855   BIC:                             2752.
Df Model:                          18                                         
Covariance Type:            nonrobust                                         
=================================================================================
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept        -0.1206      0.006    -21.050      0.000      -0.132      -0.109
cesta_compra      0.0842      0.011      7.354      0.000       0.062       0.107
bebidas_super     0.0186      0.009      2.044      0.041       0.001       0.036
bebidas_rest      0.1456      0.016      8.996      0.000       0.114       0.177
comida_rest       0.0850      0.015      5.703      0.000       0.056       0.114
ropa              0.0008      0.006      0.134      0.894      -0.011       0.013
deportes          0.0015      0.006      0.257      0.797      -0.010       0.013
cine              0.2312      0.014     16.105      0.000       0.203       0.259
marlboro          0.0410      0.008      4.904      0.000       0.025       0.057
educacion         0.0533      0.008      6.358      0.000       0.037       0.070
transporte        0.0427      0.010      4.137      0.000       0.022       0.063
gasolina          0.0230      0.007      3.183      0.001       0.009       0.037
vehiculos        -0.0147      0.005     -2.963      0.003      -0.024      -0.005
internet          0.0211      0.006      3.471      0.001       0.009       0.033
movil             0.0066      0.005      1.224      0.221      -0.004       0.017
gastos_fijos      0.0405      0.008      4.851      0.000       0.024       0.057
alquiler          0.1338      0.010     12.849      0.000       0.113       0.154
m2               -0.0080      0.004     -1.825      0.068      -0.017       0.001
inte_fij_hipo    -0.0759      0.006    -11.962      0.000      -0.088      -0.063
==============================================================================
Omnibus:                      114.312   Durbin-Watson:                   1.980
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              261.589
Skew:                          -0.027   Prob(JB):                     1.57e-57
Kurtosis:                       4.134   Cond. No.                         9.07
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

En el caso de que pudiésemos utilizar ANOVA, obtendríamos que nuestra variable respuesta "salario" estaría explicada en un 63% por las variables predictoras. Por otro lado, las variables que mejor predicen nuestra variable respuesta son "cesta_compra" con 
4889.633762 en F y "bebidas_rest" con 1603.893688 en F.